### PR TITLE
Add support for autoscaling

### PR DESCRIPTION
**Pull request by sjpb**
Most of the autoscaling setup can be done either before running runtime.yml, or using the openhpc_config variable to pass in additional slurm.conf parameters.

This PR adds an option `extra_nodes` to the group/partition definitions in openhpc_partitions to allow additional nodes definitions to be added into the slurm.conf node/partition definitions. As well as autoscaling/state=CLOUD nodes these could actually be used to add non-role-controlled normal nodes into a cluster using this role.

It also modifies the docs, as I realised they were a bit messy/confusing in places.

There are some subtleties which needed changes to the slurm.conf templating:
- "matching" inventory groups "<cluster_name>_<group_name>" won't exist if a group/partition contains only extra_nodes nodes.
- the `NodeName=DEFAULT` approach can't be used as we don't want cpu information to "fall through" to subsequent partitions which might not define all the current threads_per_core etc variables. I found existing code to group nodenames, so it uses that instead (renamed to avoid total confusion) to keep node definitions concise. As a bonus this will make the config MUCH shorter in the usual case of sequential nodenames, which will help slurmctld performance and startup time for large clusters.
- given the above changes the opportunity was taken to make the templating of node definitions clearer - generally this templating has been really hard to follow so hopefully this helps.

### CODE REVIEW

**Review by jovial — COMMENTED** [review]
Overall, change seems pretty reasonable. Templating in slurm.conf is pretty hairy, but then again it was before this change. Had a couple of little questions.

File: templates/slurm.conf.j2 [hunk 1]
```diff
@@ -106,39 +106,34 @@ NodeName={{ node }}
...
-    Nodes=\
-        {% for group in part.get('groups', [part]) %}
-            {% set group_name = group.cluster_name|default(openhpc_cluster_name) ~ '_' ~ group.name %}
-            {% if groups[group_name] | length > 0 %}{{ groups[group_name] | join(",\\\n") }}{% if not loop.last %}{{ ",\\\n" }}{% endif %}{% else %}-nonesuch{% endif %}
-        {% endfor %}{# group #}
+{% if not nodelist %}  {# empty partition - define an invalid hostname which slurm accepts #}
+    {% set nodelist = ['-nonesuch'] %}
+NodeName={{ nodelist[0] }}
+{% endif %}
+PartitionName={{part.name}} Default={{ part.get('default', 'YES') }} MaxTime={{ part.get('maxtime', openhpc_job_maxtime) }} State=UP Nodes={{ nodelist | join(',') }}
```
Comment (jovial) [inline templates/slurm.conf.j2:134]: This is actually a lot cleaner, nice...

**Review by jovial — COMMENTED** [review]
Overall, change seems pretty reasonable. Just that niggle about openhpc_config really.

**Comment by sjpb** [comment]
centos:8.2.2004, test4 failed in CI but worked ok on local molecule. Rerunning ...

**Comment by sjpb** [comment]
Passed on 2nd attempt, ready for re-review @jovial.

File: defaults/main.yml [hunk 2]
```diff
@@ -12,6 +12,8 @@ openhpc_resume_timeout: 300
 openhpc_retry_delay: 10
 openhpc_job_maxtime: 24:00:00
 openhpc_config: "{{ openhpc_extra_config | default({}) }}"
+openhpc_slurm_configless: "{{ true if ('enable_configless' in openhpc_config.get('SlurmctldParameters', [])) else false }}"
```
Comment (jovial) [inline defaults/main.yml:15]: Think this can be simplified to:
```suggestion
openhpc_slurm_configless: "{{ 'enable_configless' in openhpc_config.get('SlurmctldParameters', []) }}"
```

**Comment by sjpb** [comment]
centos:8.3.2011, test12 failed, rerunning...

**Comment by sjpb** [comment]
Passed tests on 2nd attempt

File: defaults/main.yml [hunk 3]
```diff
@@ -11,7 +11,8 @@ openhpc_drain_timeout: 86400
 openhpc_resume_timeout: 300
 openhpc_retry_delay: 10
 openhpc_job_maxtime: 24:00:00
-openhpc_config: "{{ openhpc_extra_config | default({}) }}"
+openhpc_slurm_configless: false
+openhpc_config: "{{ {'SlurmctldParameters':['enable_configless']} if openhpc_slurm_configless else {} }}"
```
[the remark this answers is not in the document]
Comment (sjpb) [reply defaults/main.yml]: As discussed, not doing that as the `_default` and `_extra` suffixes are used in the appliance. However the "configless" logic has been reversed so that `openhpc_config` is now simple.

**Comment by sjpb** [comment]
@jovial can you rereview please? Think this is ready to go.

**Review by jovial — APPROVED** [review]

File: filter_plugins/slurm_conf.py [hunk 4]
```diff
@@ -58,9 +74,16 @@ def _group_numbers(numbers):
...
+    """ Raise an error if condition is not True """
+    
+    if not condition:
+        raise errors.AnsibleFilterError(msg)
+
 class FilterModule(object):
 
     def filters(self):
         return {
-            'group_hosts': group_hosts
```
Comment (JohnGarbutt) [inline filter_plugins/slurm_conf.py:65]: That looks like its backwards incompatible? Could we keep old and new easily enough?

File: molecule/test13/verify.yml [hunk 5]
```diff
@@ -13,8 +13,8 @@
     command: scontrol show config
     register: slurm_config
   - assert:
-      that: "item in slurm_config.stdout"
+      that: "item in slurm_config.stdout_lines | map('replace', ' ', '')"
```
Comment (JohnGarbutt) [inline molecule/test13/verify.yml:16]: Might be worth adding brackets here?, I am unclear on the ordering here, I think its:
"item in (slurm_config.stdout_lines | map('replace', ' ', ''))"

File: molecule/test14/converge.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+          batch: "{{ inventory_hostname in groups['testohpc_compute'] }}"
+          runtime: true
+        openhpc_slurm_control_host: "{{ groups['testohpc_login'] | first }}"
+        openhpc_slurm_partitions:
+          - name: "compute"
+            extra_nodes:
+              # Need to specify IPs as even with State=DOWN, if slurmctld can't lookup a host it just excludes it from the config entirely
+              # Can't add to /etc/hosts via ansible due to Docker limitations on modifying /etc/hosts
+              - NodeName: fake-x,fake-y
+                NodeAddr: 0.42.42.0,0.42.42.1
```
Comment (JohnGarbutt) [inline molecule/test14/converge.yml:20]: Should we use internal IPs here? like 10.42.42.0?

Comment (JohnGarbutt) [reply molecule/test14/converge.yml:20]: Does this also affect cloud nodes that do not exist?

File: filter_plugins/slurm_conf.py (hunk 4 above)
Comment (sjpb) [reply filter_plugins/slurm_conf.py:65]: The problem is the filter name is massively confusing when used in the templating which also used `group_hosts` as a variable containing the list of hosts in that group. I can't see why backward compat is required as I can't see a use-case for another playbook using this filter.

File: molecule/test13/verify.yml (hunk 5 above)
Comment (sjpb) [reply molecule/test13/verify.yml]: Done in 505661e - testing...

File: molecule/test14/converge.yml (hunk 6 above)
Comment (sjpb) [reply molecule/test14/converge.yml:21]: @JohnGarbutt does 9b04782 clarify why I'm using invalid IPs rather than internal ones?

It doesn't affect cloud nodes which don't exist as they will have been listed in the config as State=CLOUD, not State=DOWN. The former specifically means `slurmctld` knows it doesn't know how to contact them until they're "resumed".
